ui/app/insights/src/pages/page-view.ts

```diff
@@ -0,0 +1,366 @@
+                            ></or-dashboard-preview>
+                        `, () => html`
+                            <div id="dashboard-error-text">
+                                <span>${this.getErrorMsg(true)}</span>
+                            </div>
+                        `)}
+                    `)}
+                </div>
+            </div>
+        `
+    }
+
+    getErrorMsg(translate: boolean = true): string {
+        if(this.viewDashboardOnly && this._loadedDashboards.length == 0) {
+            return (translate ? i18next.t('dashboardNotFound') : 'dashboardNotFound');
+        } else if(!this.viewDashboardOnly) {
+            return (translate ? i18next.t('noDashboardSelected-mobile') : 'noDashboardSelected-mobile')
+        }
+        return (translate ? i18next.t('errorOccured') : 'errorOccured')
+    }
+
+    selectDashboard(id: string, closeDrawer: boolean = true) {
+        if (id) {
+            if (this.dashboardMenu && closeDrawer) {
+                this.dashboardMenu.toggleDrawer(false).then(() => {
+                    this._selectedId = id;
+                });
+            } else {
+                this._selectedId = id;
+            }
+        }
+    }
+
+    // Method to switch realms.
+    // If dashboardId is not set, it will navigate to '/' since the current selectedId is not present in the new realm.
+    changeRealm(realm: string, dashboardId?: string) {
+        this._selectedId = dashboardId;
+        this.updateComplete.then(() => {
+            this._store.dispatch(updateRealm(realm));
+        })
+    }
+
+
+
+
+    /* ---------------------------- */
+
+    // PROMISE HANDLING UTILITY FUNCTIONS
+
+    // It uses _activePromises to keep track of active processes / fetches.
+    // This list is used for checking 'loading' state, or do an "await Promise.all()" to wait until everything is loaded.
+
+    isLoading(): boolean {
+        return (this._activePromises.size > 0);
+    }
+    async afterLoaded(): Promise<void> {
+        await Promise.all(this._activePromises);
+    }
+    doesPromiseExist(topic: string) {
+        return this._activePromises.has(topic);
+    }
+    getPromise(topic: string) {
+        return this._activePromises.get(topic);
+    }
+    registerPromise(topic: string, promise: Promise<any>, doUpdate: boolean = true, updateOnComplete: boolean = true): Promise<any> {
+        if(this._activePromises.has(topic)) {
+            promise = this._activePromises.get(topic);
+        } else {
+            this._activePromises.set(topic, promise);
+        }
+        promise.finally(() => { this.completePromise(topic, updateOnComplete); });
+        if(doUpdate) { this.requestUpdate("_activePromises"); }
+        return promise;
+    }
+    completePromise(topic: string, doUpdate: boolean = true) {
+        this._activePromises.delete(topic);
+        if(doUpdate) { this.requestUpdate("_activePromises"); }
+    }
```

Comment:
I haven't spent time figuring out what this custom promise code is here but maybe the `Util.Deferred` class fits your needs?

Reply:
Removed some of the unused functions in here.

Essentially it was meant to keep track of active promises for managing a "loading" state.
The problem previously that we changed the `_loading` state every time a fetch started, and when one was completed.
Like this:
```javascript
protected fetchSomething() {
    this._loading = true;
    fetch().then(() => {
        // make changes
    }).finally() => {
        this._loading = false;
    }
}
```

On this page, several HTTP fetches happen asynchronously, starting a different timestamps.
Causing `_loading` to be set to false, while other fetches are still active.
Sadly `Promise.all()` would not work here, since the promises are sometimes depending on each other,
or starting at different places in the code.

I was a bit experimental by coming up with something like this.
Having a `Map` with a `topic: string` and a `Promise` to store the active promises.
This gave flexibility of checking which fetches are active, do `Promise.all()` to wait for all promises to complete, et cetera.

But I completely understand it's a bit too much for what is necessairy haha

Open for suggestions 😄